### ISSUE TITLE
OpenGL / CUDA Interop seems completely broken

### ISSUE THREAD

**Issue by mholzel** [opencv/opencv#22546]
I am using OpenCV inside a Docker container. It is built with the following settings 
```

FROM nvidia/cudagl:11.4.1-devel-ubuntu20.04
...
ARG OPENCV_VERSION=4.6.0

RUN mkdir /opencv/ && cd /opencv/ && tar -xf /tmp/opencv-${OPENCV_VERSION}.tar.gz && tar -xf /tmp/opencv_contrib-${OPENCV_VERSION}.tar.gz && \
    mkdir /opencv/build && cd /opencv/build && \
    cmake \
    -D CMAKE_BUILD_TYPE=RELEASE \
    -D WITH_CUDA=ON \
    -D WITH_CUDA=ON \
    -D WITH_OPENGL=ON \
    -D CUDA_ARCH_BIN="6.1, 7.5, 8.0, 8.6" \
    -D BUILD_TESTS=OFF \
    -D BUILD_PERF_TESTS=OFF \
    -D BUILD_EXAMPLES=OFF \
    -D BUILD_opencv_apps=OFF \
    -D OPENCV_GENERATE_PKGCONFIG=ON \
    -D BUILD_opencv_ts=OFF \
    -D BUILD_opencv_apps=OFF \
    -D WITH_QT=ON \
    -D OpenGL_GL_PREFERENCE=GLVND \
#    -D WITH_GTK=ON \
#    -D WITH_GTK_2_X=ON \
    -D OPENCV_EXTRA_MODULES_PATH=/opencv/opencv_contrib-${OPENCV_VERSION}/modules/ \
    ../opencv-${OPENCV_VERSION}/
```

According to the configuration, QT 5 (installed via `libqt5opengl5-dev`) is correctly found and being used, along with all of the cuda functionality:
```

--   OpenCV modules:
--     To be built:                 alphamat aruco barcode bgsegm bioinspired calib3d ccalib core cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev cvv datasets dnn dnn_objdetect dnn_superres dpm face features2d flann freetype fuzzy gapi hdf hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor mcc ml objdetect optflow phase_unwrapping photo plot python3 quality rapid reg rgbd saliency shape stereo stitching structured_light superres surface_matching text tracking video videoio videostab viz wechat_qrcode xfeatures2d ximgproc xobjdetect xphoto
--     Disabled:                    world
--     Disabled by dependency:      -
--     Unavailable:                 java julia matlab ovis python2 sfm ts
--     Applications:                -
--     Documentation:               NO
--     Non-free algorithms:         NO
-- 
--   GUI:                           QT5
--     QT:                          YES (ver 5.12.8 )
--       QT OpenGL support:         YES (Qt5::OpenGL 5.12.8)
--     GTK+:                        YES (ver 2.24.32)
--       GThread :                  YES (ver 2.64.6)
--       GtkGlExt:                  YES (ver 1.2.0)
--     OpenGL support:              YES (/usr/lib/x86_64-linux-gnu/libOpenGL.so /usr/lib/x86_64-linux-gnu/libGLX.so /usr/lib/x86_64-linux-gnu/libGLU.so)
--     VTK support:                 YES (ver 7.1.1)

```

Furthermore, on their own, the Cuda and OpenGL modules seems to be fine in the sense that there are no issues building or running them. That is, there are NO issues when:

1. Converting `cv::Mat` types to `cv::cuda::GpuMat` types and vice versa. 
1. Converting `cv::Mat` types to `cv::ogl` types and vice versa. 

However, any attempt to convert between Cuda and OpenGL yields runtime errors. Here is a simple main: 

```
#include <opencv2/core/cuda.hpp>
#include <opencv2/core/opengl.hpp>
#include <opencv2/opencv.hpp>

int main() {
    //setenv("DISPLAY", ":1", true);

    cv::VideoCapture source("/tmp/opencv_opengl/topbot.png");
    cv::Mat img_cpu;
    source >> img_cpu;

    // Fine: Cuda / cv::Mat interop
    cv::cuda::GpuMat img(img_cpu);

    // Fine: OpenGL / cv::Mat interop
    cv::namedWindow("example", cv::WINDOW_OPENGL);
    cv::ogl::Texture2D img_opengl(img_cpu);

    // Nope: OpenGL / Cuda interop
    // Runtime Error:
    // OpenCV(4.6.0) /opencv/opencv-4.6.0/modules/core/src/opengl.cpp:176: error: (-217:Gpu API call) unknown error in function 'registerBuffer'
    cv::ogl::Texture2D img_opengl2;
    img_opengl2.copyFrom(img, true);

    // Nope: OpenGL / Cuda interop
    // Runtime Error:
    // OpenCV(4.6.0) /opencv/opencv-4.6.0/modules/core/src/opengl.cpp:176: error: (-217:Gpu API call) unknown error in function 'registerBuffer'
    cv::ogl::Buffer img_opengl3(100, 100, CV_8UC3);
    cv::cuda::GpuMat mapped = img_opengl3.mapDevice();

    cv::waitKey(0);
    return 0;
}

```
and an accompanying CMakeLists.txt:

```
cmake_minimum_required(VERSION 3.24)
project(opencv_opengl)

set(CMAKE_CXX_STANDARD 14)

set(OpenGL_GL_PREFERENCE GLVND)

find_package(GLUT REQUIRED)
find_package(OpenGL REQUIRED)
find_package(OpenCV 4 REQUIRED COMPONENTS
        core
        calib3d
        cudaarithm
        cudaimgproc
        cudastereo
        cudawarping
        highgui
        imgproc
        videoio
        )
find_package(Threads REQUIRED)

add_executable(opencv_opengl main.cpp)

target_link_libraries(
        opencv_opengl
        PUBLIC
        GLUT::GLUT
        OpenGL::GL
        OpenGL::GLU
        opencv_calib3d
        opencv_core
        opencv_cudaarithm
        opencv_cudaimgproc
        opencv_cudastereo
        opencv_cudawarping
        opencv_highgui
        opencv_imgproc
        opencv_videoio

)
```

To be clear, when trying to do any conversion between OpenGL and Cuda, I am hit with a runtime error:
```
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.6.0) /opencv/opencv-4.6.0/modules/core/src/opengl.cpp:176: error: (-217:Gpu API call) unknown error in function 'registerBuffer'
```

which points to this line of code: https://github.com/opencv/opencv/blob/4.6.0/modules/core/src/opengl.cpp#L176

**Comment by asmorkalov** [comment]
Does the same code work nativelly? It could be an issue of sharing hardware with docker environment.

**Comment by mholzel** [comment]
That was my thought too. I will try to test that in the next few days. Building OpenCV natively and setting everything up will just take me some time...

**Comment by mholzel** [comment]
Same error natively as in the Docker container. I built the same OpenCV version with the same parameters locally. 

Some more info... I am running an RTX A5500. The nvidia-smi output looks like this: 

`NVIDIA-SMI` 470.141.03   Driver Version: 470.141.03   CUDA Version: 11.4 `

In the past, I had issues using the cuda module that I determined to be due to the wrong CUDA_ARCH_BIN version. But now those problems seem to be solved, and the last remaining piece is to figure out why this interop is failing. It certainly seems to me that there is some function missing in the GPU driver. But on the other hand. the offending function doesn't appear to even be deprecated: 

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__OPENGL.html#group__CUDART__OPENGL_1g0fd33bea77ca7b1e69d1619caf44214b
